### PR TITLE
initial prototype for vits2

### PR DESCRIPTION
**Pull request by p0p4k**
To do some testing these are the notebooks at my other [repo](https://github.com/p0p4k/vits2_pytorch/tree/main/notebooks
).
I still have some questions about the techniques used in the paper; they do not go into the specifics. Would be great if I can get some help.

@lexkoro thanks for good discussions.

### CODE REVIEW

**Comment by p0p4k** [comment]
#2828

File: TTS/tts/layers/glow_tts/transformer.py [hunk 1]
```diff
@@ -430,3 +430,133 @@ def forward(self, x, x_mask):
             x = self.norm_layers_2[i](x + y)
         x = x * x_mask
         return x
+
+class ConditionalRelativePositionTransformer(nn.Module):
```
Comment (erogol) [inline TTS/tts/layers/glow_tts/transformer.py:434]: Better if you move this into `layers/vits2`

File: TTS/tts/layers/vits2/networks.py [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+    classname = m.__class__.__name__
+    if classname.find("Conv") != -1:
+        m.weight.data.normal_(mean, std)
+
+
+def get_padding(kernel_size, dilation=1):
+    return int((kernel_size * dilation - dilation) / 2)
+
+
+class TextEncoder(nn.Module):
```
Comment (erogol) [inline TTS/tts/layers/vits2/networks.py:29]: I guess they use regular transformer not relative position one.

**Review by erogol — COMMENTED** [review]
Looks like WIP. Pls, tag me again when you need a review.

**Comment by p0p4k** [comment]
Ahh, yes. I forgot to turn it into draft. I'll let you know as soon as it's done. Thanks!

**Comment by manmay-nakhashi** [comment]
```
    def forward_mas(self, outputs, z_p, m_p, logs_p, x, x_mask, y_mask, g, lang_emb, noise_scale=0.01):
        # find the alignment path
        attn_mask = torch.unsqueeze(x_mask, -1) * torch.unsqueeze(y_mask, 2)
        with torch.no_grad():
            o_scale = torch.exp(-2 * logs_p)
            logp1 = torch.sum(-0.5 * math.log(2 * math.pi) - logs_p, [1]).unsqueeze(-1)  # [b, t, 1]
            logp2 = torch.einsum("klm, kln -> kmn", [o_scale, -0.5 * (z_p**2)])
            logp3 = torch.einsum("klm, kln -> kmn", [m_p * o_scale, z_p])
            logp4 = torch.sum(-0.5 * (m_p**2) * o_scale, [1]).unsqueeze(-1)  # [b, t, 1]
            logp = logp2 + logp3 + logp1 + logp4
            
            # Adding Gaussian noise to the log probabilities
            epsilon = torch.randn_like(logp) * torch.std(logp) * noise_scale
            logp += epsilon
            
            attn = maximum_path(logp, attn_mask.squeeze(1)).unsqueeze(1).detach()  # [b, 1, t, t']

        # duration predictor
        attn_durations = attn.sum(3)
        if self.args.use_sdp:
            loss_duration = self.duration_predictor(
                x.detach() if self.args.detach_dp_input else x,
                x_mask,
                attn_durations,
                g=g.detach() if self.args.detach_dp_input and g is not None else g,
                lang_emb=lang_emb.detach() if self.args.detach_dp_input and lang_emb is not None else lang_emb,
            )
            loss_duration = loss_duration / torch.sum(x_mask)
        else:
            attn_log_durations = torch.log(attn_durations + 1e-6) * x_mask
            log_durations = self.duration_predictor(
                x.detach() if self.args.detach_dp_input else x,
                x_mask,
                g=g.detach() if self.args.detach_dp_input and g is not None else g,
                lang_emb=lang_emb.detach() if self.args.detach_dp_input and lang_emb is not None else lang_emb,
            )
            loss_duration = torch.sum((log_durations - attn_log_durations) ** 2, [1, 2]) / torch.sum(x_mask)
        outputs["loss_duration"] = loss_duration
        
        return outputs, attn
```
class VITS():
self.noise_decay=2e-6
self.noise_scale=0.01
 we can update noise_scale -= noise_decay in on_train_step_start. 
 @p0p4k

**Comment by erogol** [comment]
@p0p4k all looks good to me. Can you add a recipe for people to get started with training VITS2 ?

**Comment by p0p4k** [comment]
@erogol Ok, I will do some cleanup , remove unused stuff and add a recipe with test soon. Little busy with other stuff! Thanks!

**Comment by erogol** [comment]
@p0p4k sure take your time...

**Pull request closed without merging**